pagefaultgames/pokerogue
# [UI] Make Egg List and Egg Summary scrollable

## What are the changes the user will see?
- cursor wrap around in Egg List and Egg Summary 
- Egg List and Egg Summary can go past 99 and will have a scrollbar and allow scrolling if so
- Removed egg rarity background in egg summary

## Why am I making these changes?
With the addition of MEs, some of them give specific eggs as a reward which makes it possible that a player holds more than 99 eggs at the same time.
This PR makes sure the UI still works nicely in this case
The egg rarity background removal in egg summary was discussed and agreed upon a while ago as it is confusing and does not give much useful information, this was a good opportunity to do it





## What are the changes from a developer perspective?
- Created `ScrollableGridUiHandler` class that can handle all cursor update calculations for scrollable grids or lists in the game (other grids in the game will get updated to use this in a future PR)
- For Egg Summary, created `HatchedPokemonContainer` that holds the icons relevant to each hatched Pokemon instead of working with separate lists of icons
- Used the `ScrollableGridUiHandler` in both Egg Summary and Egg List
- For both of these UIs, the sprites and icons needed to fill up the grid are created only once. When scrolling happens, the sprites get updated and icons shown or hidden as needed. This avoids destroying and recreating UI elements all the time when scrolling the view




### Screenshots/Videos




Before
<img width="300" alt="_alist_before" src="https://github.com/user-attachments/assets/732244e9-179e-41e4-9f00-d8014b4b15fb"> <img width="300" alt="_asummary_before" src="https://github.com/user-attachments/assets/addc2d84-70e9-448c-9204-388bc755a8bf">

After

https://github.com/user-attachments/assets/290a2908-1c4c-4f50-9760-7f2aaf6aff53


After Egg list Default / Legacy (egg summary is the same in legacy and default theme)
<img width="300" alt="egg-list" src="https://github.com/user-attachments/assets/950597a6-7ce1-4ec3-a974-0e62251255b6"><img width="300" alt="egg-list-legacy" src="https://github.com/user-attachments/assets/d9962d69-8a6d-4ba2-8bc3-aeadb56724b0">



## How to test the changes?

[use a save file with lots of eggs](https://github.com/user-attachments/files/17101245/data_eggs.txt) or get more than 99 eggs with overrides:
```
  EGG_FREE_GACHA_PULLS_OVERRIDE: true,
  EGG_IMMEDIATE_HATCH_OVERRIDE: true,
  EGG_GACHA_PULL_COUNT_OVERRIDE: 115
```
- Play around in the egg list
- Hatch the eggs with `EGG_IMMEDIATE_HATCH_OVERRIDE` and play with the egg summary
- Make sure everything gets updated properly when moving around in various ways, especially in egg summary. Also try to pay attention to the sound effects and cries
- Also test with less than 99 eggs to make sure the scroll bar is hidden and wrap around still works, etc






## Checklist
- [ ] There is no overlap with another PR? #4370 can get closed once this gets in
- [ ] The PR is self-contained and cannot be split into smaller PRs? kinda '-'
~~[ ] If I have text, did I make it translatable and add a key in the English locale file(s)?~~
